### PR TITLE
T259233: Wikipedia Preview card dragging gestures

### PR DESCRIPTION
**Pull request by medied**
https://phabricator.wikimedia.org/T259233

# Problem

Currently, users need to tap 'Continue reading' button expand preview card. We would like to improve the user experience by allowing users to also be able to drag on mobile to interact with preview card

# Solution

Allow users to drag up to expand preview card, drag down to close it, following specified requirements:

- To drag up and reveal more content, user should be able to drag up from either the body or header of the card
- To drag down and close the Wikipedia Preview card, users should only be able to drag down the card header. This is to not interfere with any possible text scrolling
- The tap gesture should remain unchanged

Solution inspired from #56 

---

#### TODOs

- [x] Design clarification: swipe up initial touch position
- [x] Design clarification: swipe down gesture when scrolling text is present

### CODE REVIEW

**Review by hueitan — CHANGES_REQUESTED** [review]
Something missing and recommend to do it

1. Before expand, drag down should close the popup
1. After expand, does drag down more than 60px too less? too sensitive?
1. once drag match the threshold, apply the animation drag to the expanded (closed) position
1. Tap CTA (Continue reading) doesn't expand the card height
1. "throw" event that require duration calculation.

File: src/event.js [hunk 1]
```diff
@@ -64,6 +64,61 @@ export const customEvents = popup => {
...
+					const clientY = e.touches[ 0 ].clientY,
+						offset = initialY - clientY,
+						currentHeight = initialHeight + offset,
+						expanded = element.querySelector( '.wikipediapreview.expanded' )
+
+					finalY = clientY
+					if ( !isHeader && !expanded || isHeader && expanded ) {
+						containerBody.style.maxHeight = currentHeight + 'px'
+					}
+				},
```
Comment (hueitan) [inline src/event.js:92]: insert the code `e.preventDefault()` here to prevent lagging sometimes

Comment (medied) [reply src/event.js:92]: Looks like adding `e.preventDefault()` messes with the text scrolling actually

**Comment by medied** [comment]
>    Before expand, drag down should close the popup
>    After expand, does drag down more than 60px too less? too sensitive?
>    Tap CTA (Continue reading) doesn't expand the card height

Cool, addressed 1, 2 & 4

>   once drag match the threshold, apply the animation drag to the expanded (closed) position
 >  "throw" event that require duration calculation.

Could you please elaborate on 3 & 5?

**Comment by hueitan** [comment]
> Could you please elaborate on 3 & 5

> 3. once drag match the threshold, apply the animation drag to the expanded (closed) position

Currently, when the `touchend` event apply, the preview card is instantly expanded or hide, it's better to have the animation (transition in css) instead of sudden appear sudden hide

5. "throw" event that require duration calculation.

quick swipe within the given threshold, similar to gallery slider, quick swipe to get the next image instead of swiping more than 40% of the image, this gives a better experience for user who knows how to swipe.

File: src/event.js (hunk 1 above)
Comment (hueitan) [reply src/event.js:92]: Opsss, then better don't have this code.

**Comment by medied** [comment]
> > Could you please elaborate on 3 & 5
> 
> > 1. once drag match the threshold, apply the animation drag to the expanded (closed) position
> 
> Currently, when the `touchend` event apply, the preview card is instantly expanded or hide, it's better to have the animation (transition in css) instead of sudden appear sudden hide
> 
> > 1. "throw" event that require duration calculation.
> 
> quick swipe within the given threshold, similar to gallery slider, quick swipe to get the next image instead of swiping more than 40% of the image, this gives a better experience for user who knows how to swipe.

Alright I'd like to share this exploration sub-branch and have at least you and Sudhanshu play with it and see how it feels:

`T259233-swipe-more-content-animation-exploration`: https://github.com/wikimedia/wikipedia-preview/commit/24d75c208afe815bbf6ca7642c8011ee13416b0d

Some observations & thoughts: 

* The transition on dragging up feels about right to me. There's transition when the card is dragged within the threshold as well. 
* What could need some more refinement is the transition when dragging down, past the threshold to close the card
  * Since we actually switch the `visibility` to `hidden`, it still may feel a bit sudden
  * Doing some research and looking around at some solutions, what may feel better is to apply a slight "reverse animation" where the card height reduces to some point and then `visibility` gets set to `hidden`
  * But I want you all to play with it first for feedback, it may be the case as well that it feels fine the way it is right now in `T259233-swipe-more-content-animation-exploration`
* Given the current threshold, it's possible the "quick swipe/drag" is not necessary. Would like to get Sudhanshu's input on that as well.

**Review by hueitan — APPROVED** [review]
Nice gesture!

**Pull request merged**